src/Datadog.Trace.ClrProfiler.Managed/Integrations/Testing/AsyncTool.cs

```diff
@@ -0,0 +1,215 @@
+            {
+                if (previousTask.Status == TaskStatus.RanToCompletion)
+                {
+                    return Task.FromResult((TResult)continuation(((Task<TResult>)previousTask).Result, null, state));
+                }
+
+                return InternalSetTaskContinuation((Task<TResult>)previousTask, state, continuation);
+            }
+
+            public override Task SetTaskContinuationAsync<TState>(Task previousTask, TState state, Func<object, Exception, TState, Task<object>> continuation)
+                => SetTaskContinuationAsync((Task<TResult>)previousTask, state, continuation);
+
+            private static async Task<TResult> InternalSetTaskContinuation<TState>(Task<TResult> previousTask, TState state, Func<object, Exception, TState, object> continuation)
+            {
+                TResult result = default;
+                try
+                {
+                    result = await previousTask.ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    continuation(result, ex, state);
+                    throw;
+                }
+
+                return (TResult)continuation(result, null, state);
+            }
+
+            private static async Task<TResult> SetTaskContinuationAsync<TState>(Task<TResult> previousTask, TState state, Func<object, Exception, TState, Task<object>> continuation)
+            {
+                if (previousTask.Status == TaskStatus.RanToCompletion)
+                {
+                    return (TResult)await continuation(previousTask.Result, null, state).ConfigureAwait(false);
+                }
+
+                TResult result = default;
+                try
+                {
+                    result = await previousTask.ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    await continuation(result, ex, state).ConfigureAwait(false);
+                    throw;
+                }
+
+                return (TResult)await continuation(result, null, state).ConfigureAwait(false);
+            }
+        }
+
+        internal class TaskContinuationGenerator
+        {
+            public virtual Task SetTaskContinuation<TState>(Task previousTask, TState state, Func<object, Exception, TState, object> continuation)
+            {
+                if (previousTask.Status == TaskStatus.RanToCompletion)
+                {
+                    continuation(null, null, state);
+#if NET45
+                    // "If the supplied array/enumerable contains no tasks, the returned task will immediately transition to a RanToCompletion state before it's returned to the caller."
+                    // https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.whenall?redirectedfrom=MSDN&view=netframework-4.5.2#System_Threading_Tasks_Task_WhenAll_System_Threading_Tasks_Task___
+                    return Task.WhenAll();
```

Comment:
TIL. I always used something like `Task.FromResult(0)`.

Reply:
I know is not very clear, only in the docs...